File: java/src/beecrypt/crypto/DHPublicKeyImpl.java
package beecrypt.crypto;

import java.io.*;
import java.math.BigInteger;
import javax.crypto.*;
import javax.crypto.interfaces.*;
import javax.crypto.spec.*;

import beecrypt.io.*;

public class DHPublicKeyImpl implements DHPublicKey, Cloneable {
	private DHParameterSpec _params;
	private BigInteger _y;
	private byte[] _enc;

	public DHPublicKeyImpl(BigInteger p, BigInteger g, BigInteger y) {
		_params = new DHParameterSpec(p, g);
		_y = y;
	}

	public DHPublicKeyImpl(DHParameterSpec params, BigInteger y) {
		_params = params;
		_y = y;
	}

	public DHPublicKeyImpl(DHPublicKeySpec spec) {
		_params = new DHParameterSpec(spec.getP(), spec.getG());
		_y = spec.getY();
	}

	public DHPublicKeyImpl(DHPublicKey key) {
		_params = key.getParams();
		_y = key.getY();
	}

	public Object clone() {
		return new DHPublicKeyImpl(this);
	}

	public boolean equals(Object obj) {
		if (this == obj)
			return true;

		if (obj instanceof DHPublicKey) {
			DHPublicKey pri = (DHPublicKey) obj;

			if (!pri.getParams().getP().equals(_params.getP()))
				return false;
			if (!pri.getParams().getG().equals(_params.getG()))
				return false;
			if (!pri.getY().equals(_y))
				return false;

			return true;
		}
		return false;
	}

	public DHParameterSpec getParams() {
		return _params;
	}

	public BigInteger getY() {
		return _y;
	}

	public byte[] getEncoded() {
		if (_enc == null) {
			try {
				ByteArrayOutputStream bos = new ByteArrayOutputStream();
				BeeOutputStream bee = new BeeOutputStream(bos);

				bee.writeBigInteger(_params.getP());
				bee.writeBigInteger(_params.getG());
				bee.writeBigInteger(_y);
				bee.close();

				_enc = bos.toByteArray();
			} catch (IOException e) {
				// shouldn't occur
			}
		}

		return _enc;
	}

	public String getAlgorithm() {
		return "DH";
	}

	public String getFormat() {
		return "BEE";
	}
}

File: c++/provider/DHPublicKeyImpl.cxx
/*
 * Copyright (c) 2004 Beeyond Software Holding BV
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "beecrypt/c++/provider/DHPublicKeyImpl.h"
#include "beecrypt/c++/io/ByteArrayOutputStream.h"
using beecrypt::io::ByteArrayOutputStream;
#include "beecrypt/c++/beeyond/BeeOutputStream.h"
using beecrypt::beeyond::BeeOutputStream;

using namespace beecrypt::provider;

namespace {
	const String FORMAT_BEE("BEE");
	const String ALGORITHM_DH("DH");
}

DHPublicKeyImpl::DHPublicKeyImpl(const DHPublicKey& copy) : _y(copy.getY())
{
	_params = new DHParameterSpec(copy.getParams());
	_enc = 0;
}

DHPublicKeyImpl::DHPublicKeyImpl(const DHPublicKeyImpl& copy) : _y(copy._y)
{
	_params = new DHParameterSpec(*copy._params);
	_enc = 0;
}

DHPublicKeyImpl::DHPublicKeyImpl(const DHParams& params, const BigInteger& y) : _y(y)
{
	_params = new DHParameterSpec(params.getP(), params.getG(), params.getL());
	_enc = 0;
}

DHPublicKeyImpl::DHPublicKeyImpl(const dhparam& params, const mpnumber& y) : _y(y)
{
	_params = new DHParameterSpec(BigInteger(params.p), BigInteger(params.g));
	_enc = 0;
}

DHPublicKeyImpl::DHPublicKeyImpl(const BigInteger& p, const BigInteger& g, const BigInteger& y) : _y(y)
{
	_params = new DHParameterSpec(p, g);
	_enc = 0;
}

DHPublicKeyImpl::~DHPublicKeyImpl()
{
	delete _params;
	delete _enc;
}

DHPublicKeyImpl* DHPublicKeyImpl::clone() const throw ()
{
	return new DHPublicKeyImpl(*this);
}

bool DHPublicKeyImpl::equals(const Object* obj) const throw ()
{
	if (this == obj)
		return true;

	const DHPublicKey* pub = dynamic_cast<const DHPublicKey*>(obj);
	if (pub)
	{
		if (pub->getParams().getP() != _params->getP())
			return false;

		if (pub->getParams().getG() != _params->getG())
			return false;

		if (pub->getY() != _y)
			return false;

		return true;
	}

	return false;
}

const DHParams& DHPublicKeyImpl::getParams() const throw ()
{
	return *_params;
}

const BigInteger& DHPublicKeyImpl::getY() const throw ()
{
	return _y;
}

const bytearray* DHPublicKeyImpl::getEncoded() const throw ()
{
	if (!_enc)
	{
		try
		{
			ByteArrayOutputStream bos;
			BeeOutputStream bee(bos);

			bee.writeBigInteger(_params->getP());
			bee.writeBigInteger(_params->getG());
			bee.writeBigInteger(_y);
			bee.close();

			_enc = bos.toByteArray();
		}
		catch (IOException&)
		{
		}
	}

    return _enc;
}

const String& DHPublicKeyImpl::getAlgorithm() const throw ()
{
	return ALGORITHM_DH;
}

const String* DHPublicKeyImpl::getFormat() const throw ()
{
	return &FORMAT_BEE;
}
